### PR TITLE
Warn against `jax.scipy.linalg` functions in qjit and suggest using `catalyst.accelerate`

### PR DESCRIPTION
**Pull request by paul0403**
**Context:**
`catalyst.qjit` occasionally gives wrong numerical results for functions in `jax.scipy.linalg`. This is a syndrome of the more serious underlying lapack symbol isuue. See
https://app.shortcut.com/xanaduai/story/70899/find-a-system-to-automatically-create-a-custom-call-library-from-the-one-in-jax
https://github.com/PennyLaneAI/catalyst/issues/753
https://github.com/PennyLaneAI/catalyst/issues/1071
            
In the meantime, we raise a warning to recommend users to use a callback with `catalyst.accelerate` to the underlying jax function directly.

For example, instead of 
```
@qjit
def f(A):
    B = jax.scipy.linalg.expm(A)
    return B
```
 , use

```
@qjit
def f(A):
    B = catalyst.accelerate(jax.scipy.linalg.expm)(A)
    return B
```

We warn against `expm` and `LU`.

**Description of the Change:**
Detects `jax.scipy.linalg` calls in qjit and raise a warning suggesting the use of callback. 

**Benefits:**
Avoid silently wrong numerical results when `jax.scipy` functions are used inside `qjit`, e.g. `jax.scipy.linalg.expm`.

### CODE REVIEW

File: frontend/catalyst/jit.py [hunk 1]
```diff
@@ -374,6 +374,35 @@ def sum_abstracted(arr):
     if fn is None:
         return functools.partial(qjit, **kwargs)
 
+    # !!! TODO: fix jax.scipy numerical failures with properly fetched lapack calls
+    # https://app.shortcut.com/xanaduai/story/70899/find-a-system-to-automatically-create-a-custom-call-library-from-the-one-in-jax
+    # https://github.com/PennyLaneAI/catalyst/issues/753
+    # https://github.com/PennyLaneAI/catalyst/issues/1071
+    if "jax.scipy." in inspect.getsource(fn):
```
Comment (paul0403) [inline frontend/catalyst/jit.py:381]: This is a very bad way to check if the user called `jax.scipy`. Any suggestions are welcome.

Comment (paul0403) [reply frontend/catalyst/jit.py:381]: The problem is we can't change the source of `jax.scipy.linalg.expm` and raise warning over there.

**Comment by josh146** [comment]
Hey @paul0403, based on my testing there are a lot of `jax.scipy` functions that work well with qjit (and even with derivatives!). So I'm not in favour of this approach.

Could we simply patch expm/the other few functions we know don't work with accelerate?

**Comment by paul0403** [comment]
> Hey @paul0403, based on my testing there are a lot of `jax.scipy` functions that work well with qjit (and even with derivatives!). So I'm not in favour of this approach.
> 
> Could we simply patch expm/the other few functions we know don't work with accelerate?

I am working on patching.

**Comment by dime10** [comment]
Why don't we just get those functions working with the "dirty" solution we have for other functions, until we get the cleaner way of doing jax lapack callbacks figured out?

**Comment by paul0403** [comment]
> Why don't we just get those functions working with the "dirty" solution we have for other functions, until we get the cleaner way of doing jax lapack callbacks figured out?

The dirty solution sometimes give wrong numerical answers: https://github.com/PennyLaneAI/catalyst/issues/1071

File: frontend/catalyst/jit.py (hunk 1 above)
Comment (paul0403) [reply frontend/catalyst/jit.py]: no longer relevant since we patch instead of warn now

**Comment by dime10** [comment]
Ah got it, thanks for the update :)

**Comment by paul0403** [comment]
Does this need changelog? @josh146

File: frontend/catalyst/jit.py [hunk 2]
```diff
@@ -446,27 +447,33 @@ def __init__(self, fn, compile_options):
...
-        if EvaluationContext.is_tracing():
-            isQNode = isinstance(self.user_function, qml.QNode)
-            if isQNode and self.compile_options.static_argnums:
-                kwargs = {"static_argnums": self.compile_options.static_argnums, **kwargs}
+        # !!! TODO: fix jax.scipy numerical failures with properly fetched lapack calls
+        # As of now, we raise a warning prompting the user to use a callback with catalyst.accelerate()
+        # https://app.shortcut.com/xanaduai/story/70899/find-a-system-to-automatically-create-a-custom-call-library-from-the-one-in-jax
+        # https://github.com/PennyLaneAI/catalyst/issues/753
+        # https://github.com/PennyLaneAI/catalyst/issues/1071
+        with Patcher((jax.scipy.linalg, "expm", JaxLinalgWarner(jax.scipy.linalg.expm))):
```
Comment (dime10) [inline frontend/catalyst/jit.py:455]: I think this should be moved to where we perform tracing. You should some other things being patched there already.

**Comment by josh146** [comment]
> Does this need changelog? @josh146

@paul0403 yep! Here is one I prepared earlier:

````markdown
* In certain cases, `jax.scipy.linalg.expm`
  [may return incorrect numerical results](https://github.com/PennyLaneAI/catalyst/issues/1071)
  when used within a qjit-compiled function. A warning will now be raised
  when `jax.scipy.linalg.expm` is used to inform of this issue.

  In the meantime, we strongly recommend the
  [catalyst.accelerate](https://docs.pennylane.ai/projects/catalyst/en/latest/code/api/catalyst.accelerate.html) function
  within qjit-compiled function to call `jax.scipy.linalg.expm` directly.

  ```python
  @qjit
  def f(A):
      B = catalyst.accelerate(jax.scipy.linalg.expm)(A)
      return B
  ```
````

File: frontend/catalyst/jit.py (hunk 2 above)
Comment (josh146) [reply frontend/catalyst/jit.py:455]: 👍

File: frontend/catalyst/jax_extras/jax_scipy_linalg_warnings.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    For example, instead of 
+    @qjit
+    def f(A):
+        B = jax.scipy.linalg.expm(A)
+        return B
+    , use
+    @qjit
+    def f(A):
+        B = catalyst.accelerate(jax.scipy.linalg.expm)(A)
+        return B
```
Comment (josh146) [inline frontend/catalyst/jax_extras/jax_scipy_linalg_warnings.py:56]:
```suggestion
jax.scipy.linalg.expm occasionally gives wrong numerical results
when used within a qjit-compiled function.
See https://github.com/PennyLaneAI/catalyst/issues/1071.

In the meantime, we recommend catalyst.accelerate to call
the underlying expm function directly:

@qjit
def f(A):
    return catalyst.accelerate(jax.scipy.linalg.expm)(A)
```

@paul0403 made the warning more compact. I also de-indented it so it displays without indentation in the console

File: frontend/catalyst/jit.py (hunk 2 above)
Comment (paul0403) [reply frontend/catalyst/jit.py]: Good point. I put them in jit/aot compile's capture. Now after the jaxpr is produced from tracing the patches exit and doesn't linger unnecessarily for mlir.

File: frontend/catalyst/jax_extras/jax_scipy_linalg_warnings.py (hunk 3 above)
Comment (paul0403) [reply frontend/catalyst/jax_extras/jax_scipy_linalg_warnings.py:56]: should this only be for expm in the warning message? I think we should generalize it to `jax.scipy.linalg`

Comment (paul0403) [reply frontend/catalyst/jax_extras/jax_scipy_linalg_warnings.py:56]: Also, shouldn't we link to `accelerate`'s documentation?

Comment (josh146) [reply frontend/catalyst/jax_extras/jax_scipy_linalg_warnings.py:56]: discussed in person, we'll use an f-string!

Comment (paul0403) [reply frontend/catalyst/jax_extras/jax_scipy_linalg_warnings.py]: We discussed and decided to raise this warning for expm and lu.

File: frontend/catalyst/jit.py (hunk 2 above)
Comment (dime10) [reply frontend/catalyst/jit.py]: I think putting in the capture method is even better, only needs to be written once :)

**Comment by paul0403** [comment]
I tried to use `pytest.parametrize` but since the four scipy functions are slightly different there was no way to make all of them happy in a single parametrize : ( 

Not a big deal imo since these warning tests will be removed when proper lapack calls come in anyway

File: frontend/catalyst/jit.py (hunk 2 above)
Comment (paul0403) [reply frontend/catalyst/jit.py]: good point

**Review by erick-xanadu — APPROVED** [review]
Awesome! Thanks!

File: frontend/catalyst/jit.py (hunk 2 above)
Comment (dime10) [reply frontend/catalyst/jit.py]: looks good

**Review by joeycarter — APPROVED** [review]
Looks good to me as well, thanks!

**Pull request merged**